### PR TITLE
[ISSUE_137] - initial impl of PR metadata.

### PR DESCRIPTION
**Pull request by baranowb**
https://github.com/jboss-set/aphrodite/issues/137

### CODE REVIEW

File: common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java [hunk 1]
```diff
@@ -24,21 +24,25 @@
```
Comment (soul2zimate) [inline common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java]: header and copyright

File: common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java [hunk 2]
```diff
@@ -221,6 +238,179 @@ public Issue getIssue(URL url) throws NotFoundException {
...
+     * @return
+     *         <ul>
+     *         <li>null</li> - if no issue is linked in PR
+     *         <li>Issue</li> - if there is linked issue that can be fetched
+     *         </ul>
+     * @throws NotFoundException - if there is linked issue but either no tracker or issue does not exist in tracker
+     * @throws MalformedURLException
+     */
+    public Issue getIssue(final PullRequest pullRequest) throws NotFoundException, MalformedURLException {
+        Objects.requireNonNull(pullRequest, "url cannot be null");
```
Comment (soul2zimate) [inline common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java]: incorrect message "url cannot be null"

File: common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java [hunk 3]
```diff
@@ -221,6 +238,179 @@ public Issue getIssue(URL url) throws NotFoundException {
...
+     * @return
+     *         <ul>
+     *         <li>null</li> - if no issues are linked in PR
+     *         <li>Issue</li> - if there are linked issues that can be fetched
+     *         </ul>
+     * @throws NotFoundException - if there is linked issue but either no tracker or issue does not exist in tracker
+     * @throws MalformedURLException
+     */
+    public List<Issue> getRelatedIssues(final PullRequest pullRequest) throws MalformedURLException, NotFoundException {
+        Objects.requireNonNull(pullRequest, "url cannot be null");
```
Comment (soul2zimate) [inline common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java]: incorrect message "url cannot be null"

File: common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java [hunk 4]
```diff
@@ -221,6 +238,179 @@ public Issue getIssue(URL url) throws NotFoundException {
...
+     * @return
+     *         <ul>
+     *         <li>null</li> - if no issue is linked in PR or {@link #isUpstreamRequired(PullRequest)} return false;
+     *         <li>Issue</li> - if there is linked issue that can be fetched
+     *         </ul>
+     * @throws NotFoundException - if there is linked issue but either no tracker or issue does not exist in tracker
+     * @throws MalformedURLException
+     */
+    public Issue getUpstreamIssue(final PullRequest pullRequest) throws NotFoundException, MalformedURLException {
+        Objects.requireNonNull(pullRequest, "url cannot be null");
```
Comment (soul2zimate) [inline common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java]: incorrect message "url cannot be null"

File: common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java [hunk 5]
```diff
@@ -221,6 +238,179 @@ public Issue getIssue(URL url) throws NotFoundException {
...
+     * @return
+     *         <ul>
+     *         <li>null</li> - if no upstream PR is linked in PR or {@link #isUpstreamRequired(PullRequest)} return false;
+     *         <li>Issue</li> - if there is linked PR that can be fetched
+     *         </ul>
+     * @throws NotFoundException - if there is linked PR but either no tracker or issue does not exist in tracker
+     * @throws MalformedURLException
+     */
+    public PullRequest getUpstreamPullRequest(final PullRequest pullRequest) throws MalformedURLException, NotFoundException {
+        Objects.requireNonNull(pullRequest, "url cannot be null");
```
Comment (soul2zimate) [inline common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java]: incorrect message "url cannot be null"

File: common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java [hunk 6]
```diff
@@ -221,6 +238,179 @@ public Issue getIssue(URL url) throws NotFoundException {
...
+    }
+
+    /**
+     * Check if PR require upstream or not.
+     *
+     * @param pullRequest
+     * @return
+     */
+    public boolean isUpstreamRequired(final PullRequest pullRequest) {
+        Objects.requireNonNull(pullRequest, "url cannot be null");
```
Comment (soul2zimate) [inline common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java]: incorrect message "url cannot be null"

File: common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java [hunk 7]
```diff
@@ -221,6 +238,179 @@ public Issue getIssue(URL url) throws NotFoundException {
...
+        checkIssueTrackerExists();
+        final String body = pullRequest.getBody();
+        final Matcher m = UPSTREAM_ISSUE_NOT_REQUIRED.matcher(body);
+        return !m.find();
+    }
+
+    public boolean isUpgrade(PullRequest pullRequest) {
+        Objects.requireNonNull(pullRequest, "url cannot be null");
```
Comment (soul2zimate) [inline common/src/main/java/org/jboss/set/aphrodite/Aphrodite.java]: incorrect message "url cannot be null"

**Comment by baranowb** [comment]
Ive tested regex to some extent when I wrote them. APhrodite require a dedicated test suite, hence I decided to ignore it for time being, since work load is quite heavy. Yes, bad practice. No, Im not ashamed.

**Pull request merged**